### PR TITLE
Tk 157 port token tests

### PR DESCRIPTION
**Pull request by jpinsonault**
Ported the Token tests
This involved implementing various == and hash functions for Token subclasses and Config types

### CODE REVIEW

File: spec/support/shared_examples_for_object_equality.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+
+# Examples for making sure two objects are equal
```
Comment (fpringvaldsen) [inline spec/support/shared_examples_for_object_equality.rb]: Would it be possible put these examples in the `TestUtils` file? I think that may be a better place for them, as that's where they live in the original codebase.

File: spec/support/shared_examples_for_object_equality.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  it "should find the second object to be equal to the first object" do
+    expect(second_object).to eq(first_object)
+  end
+
+  it "should the hash codes of the two objects to be equal" do
+    expect(first_object.hash).to eq(second_object.hash)
+  end
+
+  it "should find the first object not equal to a random other thing" do
```
Comment (fpringvaldsen) [inline spec/support/shared_examples_for_object_equality.rb]: The two examples below are repeated across both of these shared examples, so I'm thinking maybe they should be pulled out into their own set of shared examples (I'm fairly certain you can include shared examples from another set of shared examples).

File: spec/unit/typesafe/config/token_spec.rb [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+require 'hocon'
+require 'test_utils'
+require 'pp'
+
+
+describe Hocon::Impl::Token do
+  Tokens = Hocon::Impl::Tokens
+
+  ####################
+  # Equality
```
Comment (fpringvaldsen) [inline spec/unit/typesafe/config/token_spec.rb]: I'm :+1: on this test re-organization. I think when we start getting into the more meaty and complicated test files, however, we should try to keep things as close to the ruby code as possible in terms of organization as per @cprice404's suggestion.

**Comment by fpringvaldsen** [comment]
:+1: This looks great!

File: spec/support/shared_examples_for_object_equality.rb (hunk 1 above)
Comment (jpinsonault) [inline spec/support/shared_examples_for_object_equality.rb]: @fpringvaldsen Yeah, and I think that would probably make more sense.

File: lib/hocon/impl/config_number.rb [hunk 4]
```diff
@@ -24,4 +24,29 @@ def initialize(origin, original_text)
...
+  end
+
+  def hash
+    # This hash function makes it so that a ConfigNumber with a 3.0
+    # and one with a 3 will return the hash code
+    to_int = @value.round
+
+    # If the value is an integer or a floating point equal to an integer
+    if to_int == @value
+      return to_int.hash
```
Comment (cprice404) [inline lib/hocon/impl/config_number.rb]: Well I'll be damned, I didn't even know that Ruby had this hash method.

**Pull request merged**